Support balanced parentheses at the end of DOIs

mudge opened the pull request:
During some internal data clean-up work at @altmetric, we discovered DOIs that end with closing parentheses. These were being dropped by our extraction regular expression as we enforce a word boundary (which splits when a word character is followed by a non-word character such as a closing parenthesis).

By removing the strict word boundary, we can now pick up these closing parentheses and filter out any unwanted trailing punctuation in a second pass over the matches.

We do this by explicitly dropping the last character if it is punctuation but _not_ if the DOI ends in a single, balanced parenthetical group of characters.

This means that we can successfully extract valid DOIs such as 10.1002/0471221929.ch26(v) but not invalid ones such as 10.1007/978-3-319-12580-0).

shamess committed:
Add support for capturing DOIs with trailing punctionation

The big change here is that trailing periods or quotes will also be
caught.

mudge committed:
Support balanced parentheses at the end of DOIs

During some internal data clean-up work at @altmetric, we discovered
DOIs that end with closing parentheses. These were being dropped by our
extraction regular expression as we enforce a word boundary (which
splits when a word character is followed by a non-word character such as
a closing parenthesis).

By removing the strict word boundary, we can now pick up these closing
parentheses and filter out any unwanted trailing punctuation in a second
pass over the matches.

We do this by explicitly dropping the last character if it is
punctuation but _not_ if the DOI ends in a single, balanced
parenthetical group of characters.

This means that we can successfully extract valid DOIs such as
10.1002/0471221929.ch26(v) but not invalid ones such as
10.1007/978-3-319-12580-0).

scottmatthewman commented on lib/identifiers/doi.rb:
```diff
@@ -1,7 +1,31 @@
 module Identifiers
   class DOI
+    PATTERN = %r{
```
💯 for the clarity!

cwrw replied:
second that, this is great

scottmatthewman commented on spec/identifiers/doi_spec.rb:
```diff
@@ -7,6 +7,12 @@
     expect(described_class.extract(str)).to contain_exactly('10.1049/el.2013.3006')
   end
 
+  it 'extracts DOIs from any where in a string' do
```
That should be `anywhere`, surely?

scottmatthewman commented on lib/identifiers/doi.rb:
```diff
@@ -1,7 +1,31 @@
+      str
+        .scan(PATTERN)
+        .map { |doi|
+          next doi.downcase if doi !~ /\p{Punct}$/ || doi =~ /\(.+\)$/
```
What happens if our first pass ends in multiple punctuation characters (e.g., `...`)? Wouldn't we just trim the final one, but retain the others?

scottmatthewman commented on lib/identifiers/doi.rb:
```diff
@@ -1,7 +1,31 @@
+      str
+        .scan(PATTERN)
+        .map { |doi|
+          next doi.downcase if doi !~ /\p{Punct}\z/ || doi =~ /\(.+\)\z/
```
What happens if there's more than one contiguous punctuation character after the DOI? For example, if someone uses `...` (rather than the single glyph `…`) or even gets over excited and uses `!!!`?

In these cases, do we risk only trimming the final one?

mudge replied:
We do: I can update to remove all contiguous trailing punctuation but I'll check our database to see if that will cause any unexpected issues.

mudge committed:
Remove contiguous trailing punctuation from DOIs

If DOIs are being mentioned in some larger text, they may end in
multiple contiguous punctuation (e.g. "10.1234/567.",) so strip all
trailing punctuation.

However, there is a special case for certain Wiley DOIs which end in the
string "2-#".

scottmatthewman commented on lib/identifiers/doi.rb:
```diff
@@ -22,9 +22,9 @@ def self.extract(str)
         .scan(PATTERN)
         .map { |doi|
-          next doi.downcase if doi !~ /\p{Punct}\z/ || doi =~ /\(.+\)\z/
+          next doi.downcase if doi !~ /\p{Punct}\z/ || doi =~ /\(.+\)\z/ || doi =~ /2-#\z/
```
I wonder if there's virtue in naming this new regex as a constant (e.g., `WILEY_SUFFIX`), so that the code is more self-describing?

tomstuart commented on lib/identifiers/isbn.rb:
```diff
@@ -27,7 +27,7 @@ def self.extract_ten_digit_isbns(str)
 
     def self.isbn_13_check_digit(isbn)
-      sum = isbn.each_char.zip([1, 3].cycle).reduce(0) { |sum, values| sum + (Integer(values[0]) * values[1]) }
+      sum = isbn.each_char.zip([1, 3].cycle).reduce(0) { |acc, values| acc + (Integer(values[0]) * values[1]) }
```
🤔

mudge replied:
What are you mulling? The variable shadowing fix or the line itself?

mudge replied:
#10 answers that question. I've pulled out any ISBN changes and the version bump from this PR so we can deal with those separately.
